feat(ssh_tunnel): Import/Export Databases with SSHTunnel credentials

Antonio-RiveroMartnez opened the pull request:
### SUMMARY

We are now enabling users to export databases that have SSHTunnels related to them. The exported ssh_tunnel will get included in the `database` `yaml` file and will have the used login method (`password` or `private_key + private_key_password`) with masked values.
Then when user tries to import such databases (or any dashboard/chart/dataset/savedquery), the modal will ask the user to enter such credentials.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
1. Export a DB with SSH Tunnel
2. The exported file should include the `ssh_tunnel` with masked login method
3. Try importing a DB that has `ssh_tunnel` in it
4. The modal should ask you to enter the credentials (`password` OR `private_key + private_key_password`)

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add import/export API
- Modify our import modals so users can enter required ssh_tunnel credentials that are not exported
- Update tests and create new ones

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add missing .length to if in the import modal
- Fix comment message

eschutho commented:
@Antonio-RiveroMartnez what happens when you try to import a db with ssh tunneling to an application that doesn't have the ssh tunnel feature flag turned on?

Antonio-RiveroMartnez commented:
@eschutho Thanks for reminding me about such check 👍 . It should reject the operation and show an error message to the user in such case. Let me add that validation and its tests.

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add a check for SSH_TUNNELING feature flag, if it's disabled we should raise the already known exception informing the user
- Update tests and add new tests for covering such scenario

hughhhh commented on superset/databases/schemas.py:
```diff
@@ -719,6 +723,65 @@ def validate_password(self, data: Dict[str, Any], **kwargs: Any) -> None:
+                raise SSHTunnelingNotEnabledError()
+            private_key = ssh_tunnel.get("private_key")
+            if private_key == PASSWORD_MASK:
+                raise ValidationError("Must provide a private key for the ssh tunnel")
```
can we add a test to make sure this will raise when the user is missing the private key

Antonio-RiveroMartnez replied:
Hey! I added a few in the `api_tests` and `command_tests` files:
https://github.com/apache/superset/pull/23099/files#diff-023d03cc917d159e443cf0e947dda5e28cfdf7c5a26c592860e1d419fc5b317eR651
https://github.com/apache/superset/pull/23099/files#diff-b94e273d2b41d3fe8a552cbf34cfa59ab9495154d16823b411470ae782cea38aR2473
Let me know if you consider we need to cover other scenarios in other places. Thanks

Antonio-RiveroMartnez replied:
After re-reading this, I'm adding some more checks, for example as you mentioned, not only raising when `== PASSWORD_MASK` but also checking there's at least one login method (`password` OR `private_key + private_key_password`). That way, if users manipulate files manually, we still check the integrity of the data being passed. Thanks for the insight 👍

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add more robust checks around SSH Tunnel credentials passed in the file we want to import
- Add extra exceptions so we can handle new errors and let users know
- Update and add tests for new scenarios

eschutho reviewed (approved):
LGTM!
